src/Persistence/Reflection/RuntimeReflectionProperty.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Persistence\Reflection;
+
+use Doctrine\Common\Proxy\Proxy as CommonProxy;
+use Doctrine\Persistence\Proxy;
+use ReflectionProperty;
+use ReturnTypeWillChange;
+
+use function method_exists;
+
+/**
+ * PHP Runtime Reflection Property.
+ *
+ * Avoids triggering lazy loading if the provided object
+ * is a {@see \Doctrine\Persistence\Proxy}.
+ */
+class RuntimeReflectionProperty extends ReflectionProperty
+{
+    /**
+     * {@inheritDoc}
+     *
+     * @return mixed
+     */
+    #[ReturnTypeWillChange]
+    public function getValue($object = null)
+    {
+        if ($object === null) {
+            return parent::getValue($object);
+        }
+
+        $name = $this->getName();
+        $key  = $this->isPrivate() ? "\0" . $this->class . "\0" . $name : ($this->isProtected() ? "\0*\0" . $name : $name);
+
+        return ((array) $object)[$key] ?? null;
```

Comment:
Did you do any performance or memory analysis on this compared to before? I am a bit scared :)

I never thought about the performance impact of casting an object to array before, maybe its nothing. but this will be executed hundreds to thousands of time every request where the ORM has even moderate use. Even having the getName/isPrivate/isProtected calls here in the hotpath means this change will probably have a 10%/20% hit on hydration.

This change does highlight, that instead of extending reflection we should introduce a "Mapper" abstraction for a cleaner result.

Reply:
I just did compare several strategies. In my simple test situation:
- `(array)` is the fastest alternative by 25% with one property in the class and slows down when properties are added. It compares positively to the other alternatives up to ~10 properties
- `ReflectionProperty::getValue()` is as fast as a rebound closure.

I interpret this as:
- If we consider that most entities have about ~10 properties, then `(array)` is fine
- It's not obvious to me that a data mapper would provide a significant advantage
- instead, we might want to add an API to get several properties at once - because then `(array)` will always win.

Note also this: the tests of the ORM still pass when removing this method altogether. It looks like we don't need this precaution when accessing properties. Likely because the ORM doesn't try to access properties of uninitialized proxies.

Can we remove the method then?

Reply:
Note that if this is a concern, it's already in 3.1 since #306 so that this PR doesn't change anything ;)

Reply:
> Can we remove the method then?

An idea: make using the new `RuntimeReflectionProperty` opt-in via a constructor flag, deprecate not opting in, and remove `RuntimeReflectionProperty::getValue()`.
Please advise :)

Reply:
We did some off the hand so-so benchmarks and we should be fine with this change.